cms/migrations/0017_create_and_setup_home_page.py

```diff
@@ -0,0 +1,74 @@
+# Generated by Django 2.2.10 on 2020-06-09 10:12
+
+from django.db import migrations
+from django.contrib.contenttypes.models import ContentType
+from wagtail.core.models import Page, Site
+
+from cms.models import HomePage
```

Comment:
I have done it the same way already but it does not load most of the functions. like `add_child`. If you see I have 15 commits in history that are for the same reason to show the reviewer.

Reply:
See the build was failing as long as I was importing the models from `apps`.

Reply:
Please see the commits older than it.
```try with direct model imports.  75ad48c```
The build started to pass only after import direct because most of the functions were not available.

Reply:
I was getting very strange type of errors like

- `ValueError: Cannot query "ContentType object (3)": Must be "ContentType" instance.`
- `AttributeError: 'Page' object has no attribute 'add_child`
- `AttributeError: 'StateApps' object has no attribute 'get_mode'`

Reply:
Even I have verified it once more by importing models locally from `apps`.
```
umerasghar@A006-00274  ~/arbisoft/MIT/bootcamp-ecommerce   umar/bootcamp-596 ●  docker-compose exec web python manage.py migrate cms 0016_admissions_section
Operations to perform:
  Target specific migration: 0016_admissions_section, from cms
Running migrations:
  Rendering model states... DONE
  Unapplying cms.0017_create_and_setup_home_page...Traceback (most recent call last):
  File "manage.py", line 14, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 83, in wrapped
    res = handle_func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 234, in handle
    fake_initial=fake_initial,
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 121, in migrate
    state = self._migrate_all_backwards(plan, full_plan, fake=fake)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 196, in _migrate_all_backwards
    self.unapply_migration(states[migration], migration, fake=fake)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 269, in unapply_migration
    state = migration.unapply(state, schema_editor)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/migration.py", line 175, in unapply
    operation.database_backwards(self.app_label, schema_editor, from_state, to_state)
  File "/usr/local/lib/python3.6/site-packages/django/db/migrations/operations/special.py", line 196, in database_backwards
    self.reverse_code(from_state.apps, schema_editor)
  File "/src/cms/migrations/0017_create_and_setup_home_page.py", line 66, in remove_home_page
    root.add_child(instance=default_wagtail_home_page)
AttributeError: 'Page' object has no attribute 'add_child'
 ✘ umerasghar@A006-00274  ~/arbisoft/MIT/bootcamp-ecommerce   umar/bootcamp-596 ● 
```

Reply:
Here is the old commit reference importing models from `apps` in the create and remove home page
https://github.com/mitodl/bootcamp-ecommerce/pull/631/commits/450adafdd8166dfeb4d0677d7f99706859801ec3#diff-3013b1b1421ce96a635d182d299283fcR14

Reply:
@umarmughal824 as per my knowledge, usually in migrations, we prefer to use `apps.get_model(...)` because it uses the historical versions of the model and will not complain if some field(s) in future do not exist in code base and sure then migration can complain.

Reply:
yes I know that's the preferred way but we are not getting the required functions that way.

Reply:
I just got it to work this way:

```python
def create_home_page(apps, schema_editor):
    """
    Create the home page
    """
    HomePage = apps.get_model("cms", "HomePage")

    root = Page.objects.get(depth=1)
    home_page_content_type = ContentType.objects.get_for_model(HomePage)
    valid_home_page = Page.objects.filter(
        content_type_id=home_page_content_type.id
    ).first()
    if valid_home_page is None:
        valid_home_page = HomePage(
            **DEFAULT_HOMEPAGE_PROPS,
            content_type_id=home_page_content_type.id
        )
        root.add_child(instance=valid_home_page)
        valid_home_page = Page.objects.filter(
            content_type_id=home_page_content_type.id
        ).get()
    site, _ = Site.objects.get_or_create(
        is_default_site=True,
        defaults=dict(
            root_page_id=valid_home_page.id,
            site_name=settings.WAGTAIL_SITE_NAME,
            **DEFAULT_SITE_PROPS,
        ),
    )
    if site.root_page_id != valid_home_page.id:
        site.root_page_id = valid_home_page.id
        site.save()
    wagtail_default_home_page = Page.objects.filter(
        depth=2, content_type=ContentType.objects.get_for_model(Page)
    ).first()
    # ...
```

It's a bit more code, but I would really prefer that we at least load our own models (`HomePage`) via `apps.get_model`. You can load `HomePage` in the delete function in the same way with no other changes to that function